yb-voyager/src/queryparser/traversal.go

```diff
@@ -0,0 +1,173 @@
+	if msg == nil || !msg.IsValid() {
+		return nil
+	}
+
+	if visited[msg] {
+		return nil
+	}
+	visited[msg] = true
+
+	nodeType := msg.Descriptor().FullName()
+	log.Debugf("Traversing NodeType: %s\n", nodeType)
+	// applying the processor to the current node
+	if err := processor(msg); err != nil {
+		log.Debugf("error processing node %s: %v", nodeType, err)
+		return fmt.Errorf("error processing node %s: %w", nodeType, err)
+	}
+
+	// Reference Oneof - https://protobuf.dev/programming-guides/proto3/#oneof
+	if nodeType == PG_QUERY_NODE_NODE {
+		nodeField := msg.WhichOneof(msg.Descriptor().Oneofs().ByName("node"))
+		if nodeField != nil {
+			value := msg.Get(nodeField)
+			if value.IsValid() {
+				err := TraverseParseTree(value.Message(), visited, processor)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return TraverseNodeFields(msg, visited, processor)
+}
+
+func TraverseNodeFields(msg protoreflect.Message, visited map[protoreflect.Message]bool, processor NodeProcessor) error {
+	fields := msg.Descriptor().Fields()
+	if fields == nil || fields.Len() == 0 {
+		return nil
+	}
+
+	for i := 0; i < fields.Len(); i++ {
+		fieldDesc := fields.Get(i)
+
+		// check if the field is set or not (for example: fields with DEFAULT values(set) won't be NULL)
+		if !msg.Has(fieldDesc) {
+			log.Debugf("field %q is not a part of the message ", fieldDesc.FullName())
+			continue
+		}
+
+		// (this check might be redundant due to previous Has() check)
+		// Handle oneof fields
+		if oneof := fieldDesc.ContainingOneof(); oneof != nil {
+			// Only process if the fieldDesc (field) is the one that is set in the oneof group
+			if fieldDesc != msg.WhichOneof(oneof) {
+				continue // Skip unset fields in the oneof group
+			}
+		}
+		log.Debugf("Field: %s, Type: %s\n", fieldDesc.Name(), fieldDesc.Kind())
+		value := msg.Get(fieldDesc)
+		switch {
+		case fieldDesc.IsList():
+			list := value.List()
+			for j := 0; j < list.Len(); j++ {
+				elem := list.Get(j)
+				// every elem in the list have same Kind
+				if fieldDesc.Kind() == protoreflect.MessageKind {
+					err := TraverseParseTree(elem.Message(), visited, processor)
+					if err != nil {
+						log.Debugf("error traversing field %s: %v", fieldDesc.Name(), err)
+						return fmt.Errorf("error traversing field %s: %w", fieldDesc.Name(), err)
+					}
+				}
+			}
+
+		case fieldDesc.Kind() == protoreflect.MessageKind:
+			err := TraverseParseTree(value.Message(), visited, processor)
+			if err != nil {
+				log.Debugf("error traversing field %s: %v", fieldDesc.Name(), err)
+				return fmt.Errorf("error traversing field %s: %w", fieldDesc.Name(), err)
+			}
+
+		case IsScalarKind(fieldDesc.Kind()):
+			log.Debugf("Scalar field of Type: %s, value: %v\n", fieldDesc.Kind(), value.Interface())
+		default:
+			log.Infof("field kind case not covered: %s\n", fieldDesc.Kind())
+		}
+	}
+
+	return nil
+}
+
+// ScalarKind is all datatypes including enums
```

Comment:
Is this the kind for column datatypes? If so, are we covering all the types in the list?

Reply:
Yes it will be column mostly, not sure of any special cases yet.
About coverage, yes i checked the code, apart from these only two message kinds are `Message` and `Group` for protobufs

Reply:
> If so, are we covering all the types in the list?

With this I meant the datatypes in this list

Reply:
@priyanshi-yb not sure if i get you completely.
Take a look at this - https://github.com/protocolbuffers/protobuf-go/blob/v1.35.1/reflect/protoreflect/proto.go#L248
Let me know.

Reply:
Oh okay, got it! 
The list in this functions + `Message` + `Group`  are the only  `Kind` types